src/x11.rs

```diff
@@ -1,44 +1,93 @@
+//! Implementation of software buffering for X11.
+//! 
+//! This module converts the input buffer into an XImage and then sends it over the wire to be
+//! drawn. A more effective implementation would use shared memory instead of the wire. In
+//! addition, we may also want to blit to a pixmap instead of a window.
+
 use crate::{GraphicsContextImpl, SwBufError};
 use raw_window_handle::{XlibDisplayHandle, XlibWindowHandle};
 use std::os::raw::{c_char, c_uint};
 use x11_dl::xlib::{Display, Visual, Xlib, ZPixmap, GC};
 
+/// The handle to an X11 drawing context.
 pub struct X11Impl {
+    /// The window handle.
     window_handle: XlibWindowHandle,
+
+    /// The display handle.
     display_handle: XlibDisplayHandle,
+
+    /// Reference to the X11 shared library.
     lib: Xlib,
+
+    /// The graphics context for drawing.
     gc: GC,
+
+    /// Information about the screen to use for drawing.
     visual: *mut Visual,
+
+    /// The depth (bits per pixel) of the drawing context.
     depth: i32,
 }
 
 impl X11Impl {
-    pub unsafe fn new(window_handle: XlibWindowHandle, display_handle: XlibDisplayHandle) -> Result<Self, SwBufError> {
+    /// Create a new `X11Impl` from a `XlibWindowHandle` and `XlibDisplayHandle`.
+    /// 
+    /// # Safety
+    /// 
+    /// The `XlibWindowHandle` and `XlibDisplayHandle` must be valid.
+    pub unsafe fn new(
+        window_handle: XlibWindowHandle,
+        display_handle: XlibDisplayHandle,
+    ) -> Result<Self, SwBufError> {
+        // Try to open the X11 shared library.
         let lib = match Xlib::open() {
             Ok(lib) => lib,
-            Err(e) => return Err(SwBufError::PlatformError(Some("Failed to open Xlib".into()), Some(Box::new(e))))
+            Err(e) => {
+                return Err(SwBufError::PlatformError(
+                    Some("Failed to open Xlib".into()),
+                    Some(Box::new(e)),
+                ))
+            }
         };
-        let screen = (lib.XDefaultScreen)(display_handle.display as *mut Display);
+
+        // Validate the handles to ensure that they aren't incomplete.
+        if display_handle.display.is_null() {
+            return Err(SwBufError::IncompleteDisplayHandle);
+        }
+
+        if window_handle.window.is_null() {
```

Comment:
Actually window is an integer, not a pointer, so this fails. Added a fix for this to https://github.com/rust-windowing/swbuf/pull/20.

(And CI should catch things like this in the future.)

Reply:
Yep